dualopend: reconnect during block catch-up locks in the wrong RBF inflight

If a non-tip inflight is the one that gets mined, and the peer reconnects while we are still catching up on blocks, we lock the channel in with the tip inflight -- which was never mined.

1. The block containing the mined inflight is processed. dual_funding_found fires inline and sets channel->scid, but update_channel_from_inflight runs later, from the blockdepth watches in updates_complete, only once we have caught up on all queued blocks. Until then channel->funding still points at the tip inflight.

2. The peer reconnects inside that window. peer_restart_dualopend reinits dualopend with channel_current_inflight() -- the tip -- plus channel->scid != NULL, which dualopend takes as funding-is-confirmed. It sends channel_ready.

3. The peer, who identified the mined inflight correctly, sends channel_ready back. handle_channel_locked promotes to CHANNELD_NORMAL with the tip inflight's funding and clears the inflight list, discarding the pending correction.

4. We are now in CHANNELD_NORMAL with a funding_txid that was never mined, an scid that points at the other tx's outpoint, and a last_tx that spends a nonexistent outpoint -- nothing to enforce onchain -- while the peer locked in the mined inflight, so the two sides also disagree with each other.

The window is one block-fetch wide, so any reconnect during catch-up can hit it. All it takes to set it up is an RBF replacement that failed to propagate, leaving a non-tip inflight as the one the chain picks.

We can reproduce this deterministically: stall the node's getblockhash for the block after the funding block (holding the catch-up window open) and reconnect. Fails the same way every run, in about 15 seconds. PR with the regression test and a proposed fix to follow.

----

Addendum: how it surfaced. Master run for 8c42be35c, Test CLN dual-fund Full Integration:

FAILED tests/test_opening.py::test_rbf_non_last_mined - AssertionError: assert 'd6ef2bda0490820a06461cd63d6bb75d4ba11fe2e58b9e61935d7ea904d2b3b5' == '597180eccc316cec5b817794e8f29d848b4b8364762a8e1e567f3bc7bc1e8f8a'

https://github.com/ElementsProject/lightning/actions/runs/30835065159 (logs expire, key lines below). 597180 was the mined inflight; d6ef2b was the tip, never in the mempool. l1:

17:56:23.995 lightningd: Adding block 103: 4ab240...
17:56:23.997 wallet: Owning output 0 17667576sat (p2tr) txid 597180ec... CONFIRMED
17:56:24.442 chan#1: Peer has reconnected, state DUALOPEND_AWAITING_LOCKIN: connecting subd
17:56:24.515 chan#1: State changed from DUALOPEND_AWAITING_LOCKIN to CHANNELD_NORMAL
17:56:24.602 chan#1: Funding tx d6ef2bda... depth 6 of 1
17:56:24.617 channeld-chan#1: handle_funding_depth: Setting short_channel_ids[LOCAL] to 103x1x1

(103x1x1 is 597180's funding outpoint -- its funding output is vout 1; d6ef2b's is vout 0.)

Part of #9222.

--- a/lightningd/dual_open_control.c
+++ b/lightningd/dual_open_control.c
@@ -1038,6 +1038,14 @@ static void dual_funding_found(struct lightningd *ld,
 				 loc))
 		return;
 
+	/* This inflight is the one the chain chose: record it now, so
+	 * anyone consulting the channel before we finish catching up
+	 * with the chain (e.g. a reconnecting peer) sees the mined
+	 * funding tx, not the latest RBF attempt. */
+	if (inflight->channel->state == DUALOPEND_AWAITING_LOCKIN)
+		update_channel_from_inflight(ld, inflight->channel,
+					     inflight, false);
+
 	/* Otherwise, watch for block depth increases (we'll immediately expect one) */
 	watch_blockdepth(inflight, ld->topology, loc->blkheight,
 			 opening_depth_cb,
@@ -4285,7 +4293,15 @@ bool peer_restart_dualopend(struct peer *peer,
 		       &max_to_self_delay,
 		       &min_effective_htlc_capacity);
 
-	inflight = channel_current_inflight(channel);
+	/* If a funding tx already confirmed, it is not necessarily the
+	 * latest inflight: reestablish using the one the chain chose. */
+	if (channel->scid)
+		inflight = channel_inflight_find(channel,
+						 &channel->funding.txid);
+	else
+		inflight = NULL;
+	if (!inflight)
+		inflight = channel_current_inflight(channel);
 	assert(inflight);
 	blockheight = get_blockheight(channel->blockheight_states,
 				      channel->opener, LOCAL);

--- a/tests/test_opening.py
+++ b/tests/test_opening.py
@@ -1465,6 +1465,137 @@ def test_rbf_non_last_mined(node_factory, bitcoind, chainparams):
 
 @unittest.skipIf(TEST_NETWORK != 'regtest', 'elementsd doesnt yet support PSBT features we need')
 @pytest.mark.openchannel('v2')
+def test_rbf_reconnect_non_last_mined(node_factory, bitcoind, chainparams):
+    """
+    Deterministic version of the race in test_rbf_non_last_mined.
+
+    When a non-tip RBF candidate is mined, we identify the mined
+    inflight while processing its block, but only record it on the
+    channel once we have finished catching up with the chain.  A peer
+    reconnecting inside that window reestablishes against the *newest*
+    inflight, locking the channel in with a funding tx that was never
+    mined.
+
+    We hold the window open deterministically: stall l1's fetch of the
+    block after the funding block, so l1 has seen the funding confirm
+    (its scid is set) but never finishes catching up, then reconnect.
+    """
+    l1, l2 = node_factory.get_nodes(2,
+                                    opts={'allow_warning': True,
+                                          'may_reconnect': True})
+
+    l1.rpc.connect(l2.info['id'], 'localhost', l2.port)
+    amount = 2**24
+    chan_amount = 100000
+    bitcoind.rpc.sendtoaddress(l1.rpc.newaddr()['p2tr'], amount / 10**8 + 0.01)
+    bitcoind.generate_block(1)
+    # Wait for it to arrive.
+    wait_for(lambda: len(l1.rpc.listfunds()['outputs']) > 0)
+
+    res = l1.rpc.fundchannel(l2.info['id'], chan_amount, feerate='7500perkw')
+    chan_id = res['channel_id']
+    vins = bitcoind.rpc.decoderawtransaction(res['tx'])['vin']
+    assert only_one(vins)
+    prev_utxos = ["{}:{}".format(vins[0]['txid'], vins[0]['vout'])]
+
+    # Check that we're waiting for lockin
+    l1.daemon.wait_for_log(' to DUALOPEND_AWAITING_LOCKIN')
+    inflights = only_one(l1.rpc.listpeerchannels()['channels'])['inflight']
+    assert inflights[-1]['funding_txid'] in bitcoind.rpc.getrawmempool()
+
+    def run_retry():
+        startweight = 42 + 173
+        rate = int(find_next_feerate(l1, l2)[:-5])
+        # We 2x the feerate to beat the min-relay fee
+        next_feerate = '{}perkw'.format(rate * 2)
+        initpsbt = l1.rpc.utxopsbt(chan_amount, next_feerate, startweight,
+                                   prev_utxos, reservedok=True,
+                                   excess_as_change=True)
+
+        l1.rpc.connect(l2.info['id'], 'localhost', l2.port)
+        bump = l1.rpc.openchannel_bump(chan_id, chan_amount, initpsbt['psbt'])
+        update = l1.rpc.openchannel_update(chan_id, bump['psbt'])
+        assert update['commitments_secured']
+
+        return l1.rpc.signpsbt(update['psbt'])['signed_psbt']
+
+    # Make a second inflight
+    signed_psbt = run_retry()
+    l1.rpc.openchannel_signed(chan_id, signed_psbt)
+
+    # Make it such that l1 and l2 cannot broadcast transactions
+    # (mimics failing to reach the miner with replacement)
+    def censoring_sendrawtx(r):
+        return {'id': r['id'], 'result': {}}
+
+    l1.daemon.rpcproxy.mock_rpc('sendrawtransaction', censoring_sendrawtx)
+    l2.daemon.rpcproxy.mock_rpc('sendrawtransaction', censoring_sendrawtx)
+
+    # Make a 3rd inflight that won't make it into the mempool
+    signed_psbt = run_retry()
+    last = len(l1.daemon.logs)
+    l1.rpc.openchannel_signed(chan_id, signed_psbt)
+
+    wait_for(lambda: l1.daemon.is_in_log("plugin-bcli: sendrawtx exit 0", start=last))
+    time.sleep(.05)
+
+    l1.daemon.rpcproxy.mock_rpc('sendrawtransaction', None)
+    l2.daemon.rpcproxy.mock_rpc('sendrawtransaction', None)
+
+    # We fetch out our inflights list
+    inflights = only_one(l1.rpc.listpeerchannels()['channels'])['inflight']
+    assert len(inflights) == 3
+
+    # l2 goes offline (as in the race, l1's dualopend dies with it)
+    l2.stop()
+
+    # Stall l1's fetch of the block *after* the funding block.  The
+    # mock must hang, not error: an error makes bcli report "no block
+    # yet", which lets l1 conclude it has caught up.  Returning None
+    # passes the request through to the real bitcoind.
+    height = bitcoind.rpc.getblockcount()
+    release_block = threading.Event()
+
+    def stalling_getblockhash(r):
+        if r['params'][0] == height + 2:
+            release_block.wait(timeout=180)
+        return None
+
+    l1.daemon.rpcproxy.mock_rpc('getblockhash', stalling_getblockhash)
+
+    # The 2nd inflight (the mempool tx) gets mined in block height+1.
+    bitcoind.generate_block(2, wait_for_mempool=1)
+
+    # l1 has seen the funding confirm (scid is set), but is wedged
+    # fetching block height+2, so it has not yet recorded *which*
+    # inflight was mined.
+    wait_for(lambda: 'short_channel_id'
+             in only_one(l1.rpc.listpeerchannels()['channels']))
+
+    # l2 comes back fully synced, and reconnects to wedged l1.
+    l2.start()
+    sync_blockheight(bitcoind, [l2])
+    l2.rpc.connect(l1.info['id'], 'localhost', l1.port)
+
+    l1.daemon.wait_for_log(r'to CHANNELD_NORMAL')
+    l2.daemon.wait_for_log(r'to CHANNELD_NORMAL')
+
+    # Let l1 finish catching up before we look at the result.
+    release_block.set()
+    l1.daemon.rpcproxy.mock_rpc('getblockhash', None)
+    sync_blockheight(bitcoind, [l1])
+
+    # The mined inflight (the 2nd) must be the one locked in -- on
+    # both sides, and with its commitment tx.
+    channel = only_one(l1.rpc.listpeerchannels()['channels'])
+    assert channel['funding_txid'] == inflights[1]['funding_txid']
+    assert channel['scratch_txid'] == inflights[1]['scratch_txid']
+    l2_channel = only_one(l2.rpc.listpeerchannels()['channels'])
+    assert l2_channel['funding_txid'] == inflights[1]['funding_txid']
+
+
+@unittest.skipIf(TEST_NETWORK != 'regtest', 'elementsd doesnt yet support PSBT features we need')
+@pytest.mark.openchannel('v2')
 def test_funder_options(node_factory, bitcoind):
     l1, l2, l3 = node_factory.get_nodes(3)
     l1.fundwallet(10**7)

